src/python/pants/backend/docker/parser.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import re
+from dataclasses import InitVar, dataclass, field
+from typing import Generator, Pattern
+
+from dockerfile import Command, parse_string
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.engine.fs import DigestContents, PathGlobs
+from pants.engine.rules import Get, collect_rules, rule
+from pants.option.global_options import GlobalOptions
+
+
+@dataclass(frozen=True)
+class DockerfileParseRequest:
+    sources: DockerImageSources
+
+
+@dataclass(frozen=True)
+class DockerfileInfo:
+    putative_target_addresses: tuple[str, ...] = ()
+
+
+@dataclass
+class DockerfileParser:
+    dockerfile: InitVar[str]
+    commands: tuple[Command, ...] = field(init=False)
+
+    pex_target_regexp: str = r"""
+    (?# optional path, one level with dot-separated words)
+    (?:(?P<path>(?:\w[.0-9_-]?)+) /)?
+
+    (?# binary name, with .pex file extension)
+    (?P<name>(?:\w[.0-9_-]?)+) \.pex$
+    """
+
+    def __post_init__(self, dockerfile: str):
```

Comment:
Doing real work (such as parsing) in a ctor seems weird to me. How about a `parse()` classmethod that does the parsing work and then creates the instance? Then the `@dataclass` can be `frozen=True`, which is actually necessary if this class ends up being in the signature of a `@rule` (I know it isn't at the moment, but we might as well future-proof). It's generally best to have immutable objects where possible. Or at least what passes for immutable in Python...

Reply:
Or, if we rename to parsed, how about the ctor takes the list of parsed commands as input, and have a classmethod do the parsing, returning the instance as result?
```python
@dataclass(frozen=True)
class ParsedDockerfile:
  commands: ...

  @classmethod
  def from_dockerfile(cls, dockerfile: str) -> "ParsedDockerfile":
    return cls(parse_string(dockerfile))

...
```

Reply:
Yeah, this is pretty much what I had in mind!

Reply:
Must have sped read your comment, I can see that it is what you practically spelled out :P